app/services/course_search_service.rb

```diff
@@ -36,7 +23,34 @@ def call
     scope = scope.within(filter[:radius], origin: origin) if locations_filter?
     scope = scope.with_funding_types(funding_types) if funding_types.any?
 
-    scope.distinct
+    # The 'where' scope will remove duplicates
+    # An outer query is required in the event the provider name is present.
+    # This prevents 'PG::InvalidColumnReference: ERROR: for SELECT DISTINCT, ORDER BY expressions must appear in select list'
+    outer_scope = Course.where(id: scope.pluck(:id))
+
+    if provider_name.present?
+      outer_scope = outer_scope
+                      .accredited_body_order(provider_name)
+                      .ascending_canonical_order
+    end
+
+    if sort_by_provider_ascending?
+      outer_scope = outer_scope.ascending_canonical_order
+      outer_scope = outer_scope.select("provider.provider_name", "course.*")
+    end
+
+    if sort_by_provider_descending?
+      outer_scope = outer_scope.descending_canonical_order
+      outer_scope = outer_scope.select("provider.provider_name", "course.*")
+    end
+
+    if sort_by_distance?
+      outer_scope = outer_scope.joins(sites_with_distance_from_origin)
+      outer_scope = outer_scope.select("course.*, distance")
+      outer_scope = outer_scope.order(:distance)
+    end
```

Comment:
Why not use `elsif` as before? These should mutually exclusive so it'd be nice to make that clear.